File: app/src/main/java/me/srikavin/quiz/model/Quiz.java
package me.srikavin.quiz.model;

import com.google.gson.annotations.Expose;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;


public class Quiz {
    @Expose(serialize = false)
    public String id;
    @Expose
    public String title;
    @Expose
    public String description;
    @Expose
    public List<Tag> tags;
    @Expose
    public List<QuizQuestion> questions;
    @Expose
    public boolean draft;
    @Expose
    public String overview;

    public boolean isLocal = false;

    public String coverImageUrl;

    public Quiz() {
        this.id = UUID.randomUUID().toString();
        this.tags = new ArrayList<>();
        this.description = "";
        this.questions = new ArrayList<>();
        this.draft = true;
        this.overview = "";
        this.title = "";
    }

    public Quiz(Quiz quiz) {
        this.id = quiz.id;
        this.title = quiz.title;
        this.description = quiz.description;
        this.tags = quiz.tags;
        this.questions = quiz.questions;
        this.draft = quiz.draft;
        this.overview = quiz.overview;
        this.coverImageUrl = quiz.coverImageUrl;
    }
}

File: app/src/main/java/me/srikavin/quiz/model/Quiz.kt
package me.srikavin.quiz.model

import com.google.gson.annotations.Expose
import me.srikavin.quiz.network.common.model.data.QuizModel
import me.srikavin.quiz.network.common.model.data.ResourceId
import java.util.*

/**
 * A model class representing a quiz as provided by implementations of [me.srikavin.quiz.repository.QuizRepository]
 */
data class Quiz(
        /**
         * A unique identifier used to represent a specific quiz
         */
        @Expose(serialize = false)
        override var id: ResourceId = ResourceId(UUID.randomUUID().toString()),
        /**
         * The title of the quiz
         */
        @Expose override var title: String = "",
        /**
         * A brief description of the quiz
         */
        @Expose override var description: String = "",
        /**
         * A list of tags that may contain topics or events allowing quizzes to be grouped
         */
        @Expose var tags: List<Tag> = mutableListOf(),
        /**
         * A list of questions present withing this quiz
         */
        @Expose override var questions: MutableList<QuizQuestion> = mutableListOf(),
        /**
         * Whether or not this quiz is an in-progress draft
         */
        @Expose var draft: Boolean = true,
        /**
         * A short string that may be displayed before viewing quiz details
         */
        @Expose var overview: String = "",
        /**
         * A string containing a URL of an image to display
         */
        @Expose var coverImage: String? = null,
        /**
         * The author of this quiz
         */
        @Expose(serialize = false) val author: QuizAuthor? = null,
        internal var isLocal: Boolean = false
) : QuizModel

